core/matrix/ell.hpp

```diff
@@ -0,0 +1,271 @@
+ * elements and the corresponding column indices are stored in column-major
+ * fashion. The columns are padded to the length by user-defined padding
+ * parameter whose default value is the number of rows of the matrix.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Ell : public BasicLinOp<Ell<ValueType, IndexType>>,
+            public ConvertibleTo<Dense<ValueType>>,
+            public ReadableFromMtx {
+    friend class BasicLinOp<Ell>;
+    friend class Dense<ValueType>;
+
+public:
+    using BasicLinOp<Ell>::create;
+    using BasicLinOp<Ell>::convert_to;
+    using BasicLinOp<Ell>::move_to;
+
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    void apply(const LinOp *b, LinOp *x) const override;
+
+    void apply(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+               LinOp *x) const override;
+
+    void convert_to(Dense<ValueType> *other) const override;
+
+    void move_to(Dense<ValueType> *other) override;
+
+    void read_from_mtx(const std::string &filename) override;
+
+    /**
+     * Returns the values of the matrix.
+     *
+     * @return the values of the matrix.
+     */
+    value_type *get_values() noexcept { return values_.get_data(); }
+
+    /**
+     * @copydoc Ell::get_values()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type *get_const_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * Returns the column indexes of the matrix.
+     *
+     * @return the column indexes of the matrix.
+     */
+    index_type *get_col_idxs() noexcept { return col_idxs_.get_data(); }
+
+    /**
+     * @copydoc Ell::get_col_idxs()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type *get_const_col_idxs() const noexcept
+    {
+        return col_idxs_.get_const_data();
+    }
+
+    /**
+     * Returns the maximum number of non-zeros per row.
+     *
+     * @return the maximum number of non-zeros per row.
+     */
+    size_type get_max_nonzeros_per_row() const noexcept { return max_nonzeros_per_row_; }
+
+    /**
+     * Returns the padding of the matrix.
+     *
+     * @return the padding of the matrix.
+     */
+    size_type get_padding() const noexcept { return padding_; }
+
+    /**
+     * Returns the `idx`-th non-zero element of the `row`-th row .
+     *
+     * @param row  the row of the requested element
+     * @param col  the column of the requested element
```

Comment:
I would still rename `col` to `idx` and put a different description for the parameter. If we leave it this way someone may think that `col` refers to the column in the original matrix, while it actually refers to `idx`-th non-zero in the row.

Reply:
fixed